File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/support_trains/SupportTrainReservationsStore.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.support_trains

import app.pantopus.android.data.api.models.support_trains.SupportTrainReservationDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * P3.7 — process-wide bridge between the Review-signups list and the
 * Edit-signup form. Mirrors iOS `SupportTrainReservationsStore`.
 *
 * Two responsibilities:
 *  1. Stage the seed reservation when the user taps Edit so the form
 *     can prefill without re-fetching (the backend has no single-
 *     reservation endpoint today, and `GET …/reservations` is
 *     heavy).
 *  2. Buffer optimistic save patches so the Review-signups list can
 *     replay them into its in-memory cache when the form pops.
 *
 * The backend `PATCH …/reservations/:reservationId` route lands
 * separately; until then the store's optimistic patch is the
 * user-facing source of truth — same pattern the Confirm action
 * already follows.
 */
@Singleton
class SupportTrainReservationsStore
    @Inject
    constructor() {
        private var staged: SupportTrainReservationDto? = null

        private val _revision = MutableStateFlow(0)
        /** Bumps every time a patch is applied — list VMs collect this. */
        val revision: StateFlow<Int> = _revision.asStateFlow()

        private val patches: MutableMap<String, SupportTrainReservationDto> = mutableMapOf()

        /** Stash the row the user just tapped Edit on. */
        fun stage(reservation: SupportTrainReservationDto) {
            staged = reservation
        }

        /** Take — and clear — the staged seed. Form VMs call this on init. */
        fun consumeStaged(): SupportTrainReservationDto? {
            val out = staged
            staged = null
            return out
        }

        /** Record an optimistic save patch. Bumps [revision] so observers re-render. */
        fun applyPatch(updated: SupportTrainReservationDto) {
            patches[updated.id] = updated
            _revision.value = _revision.value + 1
        }

        /**
         * Return — and remove — the patch for [reservationId], or null
         * when there is no pending edit. The Review-signups VM drains
         * patches into its in-memory cache after the form pops.
         */
        fun consumePatch(reservationId: String): SupportTrainReservationDto? = patches.remove(reservationId)

        /** Test affordance — empty the store between unit tests. */
        fun reset() {
            staged = null
            patches.clear()
            _revision.value = 0
        }
    }

File: frontend/apps/ios/Pantopus/Features/SupportTrains/SupportTrainReservationsStore.swift
//
//  SupportTrainReservationsStore.swift
//  Pantopus
//
//  P3.7 — Bridge between the Edit Signup form and the Review Signups
//  list. The form pushes the updated reservation here when the
//  organizer hits Save; the list view-model pulls patches on appear
//  and replays them against its in-memory cache so the row reflects
//  the change without a re-fetch (the backend `PATCH …/reservations
//  /:reservationId` endpoint lands separately — until then this is
//  the user-facing source of truth, matching the optimistic-confirm
//  pattern already wired in `ReviewSignupsViewModel.confirm(_:)`).
//

import Foundation
import Observation

/// Process-wide cache of optimistic reservation patches keyed by
/// reservation id. The list screen reads from here on appear; the
/// edit form writes here on Save.
@Observable
@MainActor
public final class SupportTrainReservationsStore {
    public static let shared = SupportTrainReservationsStore()

    /// Bumps every time a patch is applied — the Review-signups view
    /// observes this in `.onChange` to refresh its rows.
    public private(set) var revision: Int = 0

    private var patches: [String: SupportTrainReservationDTO] = [:]

    /// Record a patched reservation. Increments `revision` so observers
    /// re-render.
    public func apply(_ updated: SupportTrainReservationDTO) {
        patches[updated.id] = updated
        revision &+= 1
    }

    /// Return — and remove — the patch for `reservationId`, or `nil`
    /// when there is no pending edit. Called by the list VM after it
    /// has replayed the patch into its own cache so subsequent appears
    /// don't double-apply.
    public func consumePatch(forId reservationId: String) -> SupportTrainReservationDTO? {
        patches.removeValue(forKey: reservationId)
    }

    /// Test affordance — empty the store between unit tests.
    public func reset() {
        patches.removeAll()
        revision = 0
    }

    private init() {}
}
